(DOCSP-26795): Flutter manage sync session/subscriptions refactor (#2355)
mongodb/docs-realm: source/sdk/flutter/sync/manage-sync-session.txt

   :depth: 2
   :class: singlecol

When you use Atlas Device Sync, the Realm Flutter SDK syncs data with Atlas
in the background using a sync session. The sync session starts whenever
you open a synced realm.

The sync session manages the following:

- Uploading and downloading changes to the realm
- Pausing and resuming sync
- Monitoring sync progress
- Monitoring network connectivity

You can access the :flutter-sdk:`Session <realm/Session-class.html>` of any synced
realm through the :flutter-sdk:`Realm.syncSession <realm/Realm/syncSession.html>` property.

Prerequisites
-------------

Before you can manage your sync session state, you must perform the following:

#. :ref:`Configure Flexible Sync on the Atlas App Services backend <enable-flexible-sync>`
#. :ref:`Add Device Sync to your app <flutter-add-sync-to-app>`

.. _flutter-sync-wait-for-changes:

Wait for Changes to Upload and Download
---------------------------------------

> To asynchronously wait for your local changes to upload to Atlas,

Author: back to you! see changes

call the :flutter-sdk:`Session.waitForUpload()
<realm/Session/waitForUpload.html>`. To asynchronously wait for all changes on Atlas
to download to your synced realm, call :flutter-sdk:`Session.waitForDownload()
<realm/Session/waitForDownload.html>`.

.. literalinclude:: /examples/generated/flutter/manage_sync_session_test.snippet.wait-upload-download.dart
   :language: dart

.. _flutter-pause-resume-sync:

Pause and Resume a Sync Session

> -------------------------------

Author: interesting thought, but i don't think there's any single best practice here!

To pause syncing for a session, call :flutter-sdk:`Session.pause() <realm/Session/pause.html>`.
The realm will not sync changes with Atlas while the session is paused.

To resume syncing a changes, call :flutter-sdk:`Session.resume() <realm/Session/resume.html>`.

You must manually call ``Session.pause()`` and ``Session.resume()`` for each
realm whose Sync session you want to pause and restart.
The sync state of one session has no impact on other open sessions.

The following code block demonstrates calling these methods:

.. literalinclude:: /examples/generated/flutter/manage_sync_session_test.snippet.pause-resume-sync.dart
   :language: dart

.. _flutter-monitor-sync-progress:

Monitor Sync Progress
---------------------

To monitor Sync progress, call :flutter-sdk:`SyncSession.getProgressStream()

> <realm/Session/getProgressStream.html>`. This method returns a Stream of

Author: yes, it's a Dart data type - https://api.dart.dev/stable/2.18.5/dart-async/Stream-class.html

it's a common data type in the dart/flutter world, and i think would make sense to developers as such. (and it's a common enough type that i don't think need to link to those reference docs here)

> <realm/Session/getProgressStream.html>`. This method returns a Stream of

Reviewer: Makes sense. My ignorance of of dart/flutter lead to the question. Thanks for educating me.

   :language: dart

Considerations when monitoring sync progress with ``SyncSession.getProgressStream()``:

- The number of transferred and transferable bytes are only estimates.
  The Sync changesets are compressed with
  `gzip <https://www.gnu.org/software/gzip/>`_ before transmitting, so
  the actual size of transmitted bytes will be smaller than the reported number
  of both transferable and transferred bytes.
- Realm optimizes download speeds by combining multiple changesets
  into a single download message, up to 16 MB. Since the progress callback is
  only invoked once before and after a download message is processed, this
  means that you'll likely see ``transferredBytes`` change in increments of roughly
  16 MB rather than continuously as the message is being downloaded.

.. _flutter-monitor-network-connection:

Monitor Network Connection
--------------------------

You can get the state of the current network connection with
:flutter-sdk:`Session.connectionState <realm/Session/connectionState.html>`.
This returns a :flutter-sdk:`ConnectionState <realm/ConnectionState.html>` enum

> with whether the network connection is ``connected``, ``disconnected``, ``connecting``.

Reviewer: Minor wording suggestion.
```suggestion
that contains the network connection's state: ``connected``, ``disconnected``, or ``connecting``.
```

.. literalinclude:: /examples/generated/flutter/manage_sync_session_test.snippet.get-network-connection.dart
   :language: dart

Monitor the state of the network connection with
:flutter-sdk:`Session.connectionStateChanges <realm/Session/connectionStateChanges.html>`.

> This property returns a Stream of :flutter-sdk:`ConnectionStateChange <realm/ConnectionStateChange-class.html>`

Reviewer: Lowercase stream.
```suggestion
This property returns a stream of :flutter-sdk:`ConnectionStateChange <realm/ConnectionStateChange-class.html>`
```
